File: bot/threads.py
from __future__ import annotations

from typing import Any, Literal
from urllib.parse import quote

from bot.config import ThreadsSettings
from bot.http import JsonHttpClient

SearchType = Literal["TOP", "RECENT"]
SearchMode = Literal["KEYWORD", "TAG"]


class ThreadsClient:
    def __init__(self, settings: ThreadsSettings, http: JsonHttpClient | None = None) -> None:
        self.base_url = settings.api_base_url.rstrip("/")
        self.user_id = quote(settings.user_id, safe="")
        self.http = http or JsonHttpClient()
        self.headers = {"Authorization": f"Bearer {settings.access_token}"}

    def create_container(
        self,
        text: str,
        *,
        media_url: str | None = None,
        reply_to_id: str | None = None,
    ) -> str:
        query: dict[str, str] = {"text": text}
        if media_url:
            lowercase_url = media_url.lower().split("?", 1)[0]
            if lowercase_url.endswith((".mp4", ".mov", ".webm")):
                query.update({"media_type": "VIDEO", "video_url": media_url})
            else:
                query.update({"media_type": "IMAGE", "image_url": media_url})
        else:
            query["media_type"] = "TEXT"
        if reply_to_id:
            query["reply_to_id"] = reply_to_id

        data = self.http.request(
            "POST",
            f"{self.base_url}/{self.user_id}/threads",
            headers=self.headers,
            query=query,
        )
        if not isinstance(data, dict) or not data.get("id"):
            raise RuntimeError("Threads create container response has no id")
        return str(data["id"])

    def publish_container(self, container_id: str) -> str:
        data = self.http.request(
            "POST",
            f"{self.base_url}/{self.user_id}/threads_publish",
            headers=self.headers,
            query={"creation_id": container_id},
        )
        if not isinstance(data, dict) or not data.get("id"):
            raise RuntimeError("Threads publish response has no id")
        return str(data["id"])

    def reply_to(self, reply_id: str, text: str) -> str:
        container_id = self.create_container(text, reply_to_id=reply_id)
        return self.publish_container(container_id)

    def keyword_search(
        self,
        query_text: str,
        *,
        search_type: SearchType,
        search_mode: SearchMode,
        limit: int = 25,
    ) -> list[dict[str, Any]]:
        data = self.http.request(
            "GET",
            f"{self.base_url}/keyword_search",
            headers=self.headers,
            query={
                "q": query_text,
                "search_type": search_type,
                "search_mode": search_mode,
                "fields": "id,text,username,permalink,timestamp",
                "limit": max(1, min(limit, 50)),
            },
        )
        if not isinstance(data, dict) or not isinstance(data.get("data"), list):
            raise RuntimeError("Threads keyword search returned an unexpected response")
        return [item for item in data["data"] if isinstance(item, dict)]

File: supabase/functions/_shared/threads.ts
import { fetchJson } from "./http.ts";

interface IdResponse {
  id?: string;
}

export interface ThreadsSearchPost {
  id?: string;
  text?: string;
  username?: string;
  permalink?: string;
  timestamp?: string;
}

export interface ThreadsOwnPost {
  id?: string;
  timestamp?: string;
  has_replies?: boolean;
}

export interface ThreadsReply {
  id?: string;
  text?: string;
  username?: string;
  timestamp?: string;
  is_reply_owned_by_me?: boolean;
}

export interface ThreadsMention {
  id?: string;
  text?: string;
  username?: string;
  permalink?: string;
  timestamp?: string;
}

interface ThreadsListResponse<T> {
  data?: T[];
}

export class ThreadsClient {
  constructor(
    private readonly accessToken: string,
    private readonly userId: string,
    private readonly baseUrl = "https://graph.threads.net",
  ) {}

  private url(path: string, parameters: Record<string, string>): URL {
    const url = new URL(`${this.baseUrl.replace(/\/$/, "")}/${path.replace(/^\//, "")}`);
    for (const [key, value] of Object.entries(parameters)) {
      url.searchParams.set(key, value);
    }
    return url;
  }

  private requestInit(method: "GET" | "POST"): RequestInit {
    return {
      method,
      headers: { authorization: `Bearer ${this.accessToken}` },
    };
  }

  async createContainer(
    text: string,
    options: { mediaUrl?: string | null; replyToId?: string } = {},
  ): Promise<string> {
    let mediaType = "TEXT";
    const parameters: Record<string, string> = { text };
    if (options.mediaUrl) {
      mediaType = /\.(mp4|mov|webm)(?:$|\?)/i.test(options.mediaUrl) ? "VIDEO" : "IMAGE";
      parameters[mediaType === "VIDEO" ? "video_url" : "image_url"] = options.mediaUrl;
    }
    parameters.media_type = mediaType;
    if (options.replyToId) parameters.reply_to_id = options.replyToId;

    const payload = await fetchJson<IdResponse>(
      "Threads API",
      this.url(`${this.userId}/threads`, parameters),
      this.requestInit("POST"),
    );
    if (!payload.id) throw new Error("Threads API returned no container id");
    return payload.id;
  }

  async publishContainer(containerId: string): Promise<string> {
    const payload = await fetchJson<IdResponse>(
      "Threads API",
      this.url(`${this.userId}/threads_publish`, { creation_id: containerId }),
      this.requestInit("POST"),
    );
    if (!payload.id) throw new Error("Threads API returned no published post id");
    return payload.id;
  }

  async reply(replyToId: string, text: string): Promise<string> {
    const containerId = await this.createContainer(text, { replyToId });
    return this.publishContainer(containerId);
  }

  async keywordSearch(
    query: string,
    searchType: "TOP" | "RECENT",
    searchMode: "KEYWORD" | "TAG",
    limit = 25,
  ): Promise<ThreadsSearchPost[]> {
    const payload = await fetchJson<ThreadsListResponse<ThreadsSearchPost>>(
      "Threads API",
      this.url("keyword_search", {
        q: query,
        search_type: searchType,
        search_mode: searchMode,
        fields: "id,text,username,permalink,timestamp",
        limit: String(limit),
      }),
      this.requestInit("GET"),
    );
    return Array.isArray(payload.data) ? payload.data : [];
  }

  async ownPosts(limit = 5): Promise<ThreadsOwnPost[]> {
    const payload = await fetchJson<ThreadsListResponse<ThreadsOwnPost>>(
      "Threads API",
      this.url("me/threads", {
        fields: "id,timestamp,has_replies",
        limit: String(limit),
      }),
      this.requestInit("GET"),
    );
    return Array.isArray(payload.data) ? payload.data : [];
  }

  async replies(threadId: string, limit = 50): Promise<ThreadsReply[]> {
    const payload = await fetchJson<ThreadsListResponse<ThreadsReply>>(
      "Threads API",
      this.url(`${threadId}/replies`, {
        fields: "id,text,username,timestamp,is_reply_owned_by_me",
        reverse: "true",
        limit: String(limit),
      }),
      this.requestInit("GET"),
    );
    return Array.isArray(payload.data) ? payload.data : [];
  }

  async mentions(limit = 50): Promise<ThreadsMention[]> {
    const payload = await fetchJson<ThreadsListResponse<ThreadsMention>>(
      "Threads API",
      this.url("me/mentions", {
        fields: "id,text,username,permalink,timestamp",
        limit: String(limit),
      }),
      this.requestInit("GET"),
    );
    return Array.isArray(payload.data) ? payload.data : [];
  }
}
